### PR TITLE
Improve supertrend strategy

### PR DESCRIPTION
**Pull request by andycheng123**
```
winningRatio: "1.50000000"
numOfLossTrade: 4
numOfProfitTrade: 6
grossProfit: "9510.58490673"
grossLoss: "-1616.37687265"
profits:
    - "962.02614650"
    - "1492.98331752"
    - "2013.07243890"
    - "0.00006635"
    - "557.31764608"
    - "4485.18529138"
losses:
    - "-402.85689998"
    - "-384.84449514"
    - "-376.47721820"
    - "-452.19825933"
mostProfitableTrade: "4485.18529138"
mostLossTrade: "-452.19825933"
profitFactor: "-5.88389073"
totalNetProfit: "7894.20803408"

BACK-TEST REPORT
===============================================
START TIME: Tue, 01 Mar 2022 00:00:00 UTC
END TIME: Thu, 30 Jun 2022 00:00:00 UTC
INITIAL TOTAL BALANCE: BalanceMap[BTC: 1, USDT: 10000]
FINAL TOTAL BALANCE: BalanceMap[BTC: 1.885876, USDT: 0.0000659]
binance BTCUSDT PROFIT AND LOSS REPORT
===============================================
TRADES SINCE: 2022-03-01 21:04:59.999 +0000 UTC
NUMBER OF TRADES: 19
AVERAGE COST: $ 20,199.45
BASE ASSET POSITION: 0.885876
TOTAL BUY VOLUME: 3.49511477
TOTAL SELL VOLUME: 2.60923877
CURRENT PRICE: $ 20,122.58
CURRENCY FEES:
 - FEE: 152.2463837
PROFIT: $ 7,894.21
UNREALIZED PROFIT: -$ 68.10
INITIAL ASSET IN USDT ~= 53160.00000 USDT (1 BTC = 43160)
FINAL ASSET IN USDT ~= 37948.69074 USDT (1 BTC = 20122.58)
REALIZED PROFIT: +7894.20803408 USDT
UNREALIZED PROFIT: -68.09728812 USDT
ASSET DECREASED: -15211.30925402 USDT (-28.61%)
```

### CODE REVIEW

File: pkg/strategy/supertrend/strategy.go [hunk 1]
```diff
@@ -156,16 +216,34 @@ func (s *Strategy) ClosePosition(ctx context.Context, percentage fixedpoint.Valu
...
 	}
 	s.slowDEMA = &indicator.DEMA{IntervalWindow: types.IntervalWindow{Interval: s.Interval, Window: s.SlowDEMAWindow}}
+	s.slowDEMA.Bind(kLineStore)
+	// Preload
+	if klines, ok := kLineStore.KLinesOfInterval(s.fastDEMA.Interval); ok {
+		for i := 0; i < len(*klines); i++ {
+			s.fastDEMA.Update((*klines)[i].GetClose().Float64())
+		}
+	}
+	if klines, ok := kLineStore.KLinesOfInterval(s.slowDEMA.Interval); ok {
```
Comment (c9s) [inline pkg/strategy/supertrend/strategy.go:240]: pull out this preload code to a function?

File: pkg/strategy/supertrend/strategy.go [hunk 2]
```diff
@@ -174,22 +252,26 @@ func (s *Strategy) setupIndicators() {
 	}
 	s.Supertrend = &indicator.Supertrend{IntervalWindow: types.IntervalWindow{Window: s.SupertrendWindow, Interval: s.Interval}, ATRMultiplier: s.SupertrendMultiplier}
 	s.Supertrend.AverageTrueRange = &indicator.ATR{IntervalWindow: types.IntervalWindow{Window: s.SupertrendWindow, Interval: s.Interval}}
+	s.Supertrend.Bind(kLineStore)
+	// Preload
+	if klines, ok := kLineStore.KLinesOfInterval(s.Supertrend.Interval); ok {
```
Comment (c9s) [inline pkg/strategy/supertrend/strategy.go:257]: ditto.

File: pkg/strategy/supertrend/strategy.go [hunk 3]
```diff
@@ -309,72 +402,85 @@ func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, se
...
+		if s.LinearRegression != nil {
+			switch {
+			case s.LinearRegression.baseLineSlope > 0:
+				lgSignal = types.DirectionUp
+			case s.LinearRegression.baseLineSlope < 0:
+				lgSignal = types.DirectionDown
+			default:
+				lgSignal = types.DirectionNone
+			}
+		}
```
Comment (c9s) [inline pkg/strategy/supertrend/strategy.go:416]: can you pull this out to a function?

File: pkg/strategy/supertrend/strategy.go [hunk 4]
```diff
@@ -22,13 +23,60 @@ const stateKey = "state-v1"
...
+
 func init() {
 	// Register the pointer of the strategy struct,
 	// so that bbgo knows what struct to be used to unmarshal the configs (YAML or JSON)
 	// Note: built-in strategies need to imported manually in the bbgo cmd package.
 	bbgo.RegisterStrategy(ID, &Strategy{})
 }
 
+// LinGre is Linear Regression baseline
+type LinGre struct {
```
Comment (c9s) [inline pkg/strategy/supertrend/strategy.go:36]: Move this to a single file?

File: pkg/strategy/supertrend/strategy.go [hunk 5]
```diff
@@ -70,21 +119,33 @@ type Strategy struct {
...
+	StopByReversedDema bool `json:"stopByReversedDema"`
+
+	// StopByReversedLinGre TP/SL by reversed linear regression signal
+	StopByReversedLinGre bool `json:"stopByReversedLinGre"`
 
 	currentTakeProfitPrice fixedpoint.Value
 	currentStopLossPrice   fixedpoint.Value
 
+	// ExitMethods Exit methods
+	ExitMethods []bbgo.ExitMethod `json:"exits"`
```
Comment (c9s) [inline pkg/strategy/supertrend/strategy.go:147]: use bbgo.ExitMethodSet

File: pkg/strategy/supertrend/strategy.go (hunk 4 above)
Comment (andycheng123) [reply pkg/strategy/supertrend/strategy.go:36]: It's not enough to be a standalone indicator
should it be in a single file?

File: pkg/strategy/supertrend/strategy.go (hunk 5 above)
Comment (andycheng123) [reply pkg/strategy/supertrend/strategy.go:147]: ExitMethodSet is not in this branch yet
replace it later?

File: config/supertrend.yaml [hunk 6]
```diff
@@ -18,8 +18,8 @@ backtest:
   # for testing max draw down (MDD) at 03-12
   # see here for more details
   # https://www.investopedia.com/terms/m/maximum-drawdown-mdd.asp
-  startTime: "2022-04-01"
-  endTime: "2022-04-30"
+  startTime: "2022-03-01"
```
Comment (c9s) [inline config/supertrend.yaml:21]: let's always start from 2022-01-01 ?

**Review by c9s — APPROVED** [review]

**Pull request merged**